File: lintcode/python/0155_minimum_depth_of_binary_tree.py
#!/usr/bin/python -t

# dfs

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    """
    @param root: The root of binary tree
    @return: An integer
    """
    def minDepth(self, root):
        # write your code here
        if not root:
            return 0
            
        return self.dfs(root)
           
    def dfs(self, root):
        if not root:
            return sys.maxint
            
        if not root.left and not root.right:
            return 1
        
        min_left = self.dfs(root.left)
        min_right = self.dfs(root.right)
        
        return min(min_left, min_right) + 1

File: lintcode/golang/0155_minimum_depth_of_binary_tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

/**
 * @param root: The root of binary tree
 * @return: An integer
 */
func minDepth (root *TreeNode) int {
    // write your code here
    var ret int
    
    if root == nil {
        return ret
    }
    
    q := make([]*TreeNode, 0)
    q = append(q, root)
    ret++
    
    for len(q) != 0 {
        level := make([]*TreeNode, 0)
        for _, node := range q {
            if node.Left == nil && node.Right == nil {
                return ret
            }
            if node.Left != nil {
                level = append(level, node.Left)
            }
            if node.Right != nil {
                level = append(level, node.Right)
            }
        }
        ret++
        q = level
    }
    
    return ret
}

